### PR TITLE
CCO-353: ccoctl to create azure custom roles

### PR DESCRIPTION
**Pull request by jstuever**
ccoctl azure to create custom roles for each credential request which contains permissions.

When ccoctl azure operates on a credential request which has the new permissions definition, a new custom role will be created for the operator with the specified permissions. This custom role will be added to the list of roles specified, if any, and a role assignment will be created for the associated managed identity.

https://issues.redhat.com/browse/CCO-353

### CODE REVIEW

**Comment by jstuever** [comment]
There is currently an issue where sometimes an error is thrown when the new custom role is not fully propagated. Continue work to resolve this. A workaround is to rerun the command multiple times while waiting 5 minutes between each run (to avoid creating duplicate roles).

File: pkg/cmd/provisioning/azure/create_managed_identities_test.go [hunk 1]
```diff
@@ -118,8 +137,8 @@ func TestCreateManagedIdentities(t *testing.T) {
 				wrapper := mockAzureClientWrapper(mockCtrl)
 				mockGetResourceGroupNotFound(wrapper, testInstallResourceGroupName, testSubscriptionID)
 				mockCreateOrUpdateResourceGroupSuccess(wrapper, testInstallResourceGroupName, testRegionName, testSubscriptionID, resourceTags)
-				mockGetUserAssignedManagedIdentityNotFound(wrapper, testOIDCResourceGroupName, "testinfraname-secretName1-namespace1")
-				mockCreateOrUpdateManagedIdentitySuccess(wrapper, testOIDCResourceGroupName, "testinfraname-secretName1-namespace1", testRegionName, testSubscriptionID, resourceTags)
+				mockGetUserAssignedManagedIdentityNotFound(wrapper, testOIDCResourceGroupName, "testinfraname-namespace1-secretName1")
```
Comment (abutcher) [inline pkg/cmd/provisioning/azure/create_managed_identities_test.go:140]: Nice catch, I must have reversed these?

**Review by abutcher — COMMENTED** [review]
Looks straightforward! I will give this a try.

File: pkg/apis/cloudcredential/v1/types_azure.go [hunk 2]
```diff
@@ -31,6 +31,13 @@ type AzureProviderSpec struct {
 
 	// RoleBindings contains a list of roles that should be associated with the minted credential.
 	RoleBindings []RoleBinding `json:"roleBindings"`
+
+	// Permissions is the list of Azure permissions required to create a more fine-grained custom role to
+	// satisfy the CredentialsRequest.
+	// The Permissions field may be provided in addition to RoleBindings. When both fields are specified,
+	// the service account will have union of permissions defined from both Permissions and RoleBindings.
```
[the remark this answers is not in the document]
Comment (abutcher) [reply pkg/apis/cloudcredential/v1/types_azure.go:38]: Suggested change is to use "user-assigned managed identity" in place of "service account". I see latent comments above regarding minted credentials but we can fix those separately. Azure credentials aren't minted any more aside from managed identities.

[the remark this answers is not in the document]
Comment (jstuever) [reply pkg/apis/cloudcredential/v1/types_azure.go]: [Done](https://github.com/openshift/cloud-credential-operator/compare/6036ac683b09d4d1d5b23ae66923d42da7315d2e..ca18fc7903a7ec5a5d659a6a9688021b0b42c52e)

File: pkg/cmd/provisioning/azure/create_managed_identities_test.go (hunk 1 above)
Comment (jstuever) [reply pkg/cmd/provisioning/azure/create_managed_identities_test.go:140]: The [templates](https://github.com/openshift/cloud-credential-operator/pull/556/files#diff-c25bef353d0f432f8ba848ebbe66e75a45d79b2261eb7d4932fa0857a1fa3bb7R80-R81) were in order of 1) `name`, 2) `namespace`, but the template [usage](https://github.com/openshift/cloud-credential-operator/pull/556/files#diff-c25bef353d0f432f8ba848ebbe66e75a45d79b2261eb7d4932fa0857a1fa3bb7R729-R735) was feeding `namespace` before `name`. So it was reversed there, and all of the mock usage was reversed to match.

**Comment by abutcher** [comment]
In testing I was hitting replication delays which @jstuever addressed in [ca18fc..82963b](https://github.com/openshift/cloud-credential-operator/compare/ca18fc7903a7ec5a5d659a6a9688021b0b42c52e..82963bce31c5b8de315c40918c0112d439757594)

```
2023/07/07 15:53:31 Failed to get role definition, retrying...
2023/07/07 15:53:42 Failed to get role definition, retrying...
2023/07/07 15:53:52 Failed to get role definition, retrying...
2023/07/07 15:54:02 Failed to get role definition, retrying...
2023/07/07 15:54:12 Failed to get role definition, retrying...
2023/07/07 15:54:22 Failed to get role definition, retrying...
2023/07/07 15:54:33 failed to get role definition for role abutcherdemo-openshift-ingress-operator-cloud-credentials: no role found for 
name abutcherdemo-openshift-ingress-operator-cloud-credentials
```

```
2023/07/07 15:54:49 error ensuring custom role: PUT https://management.azure.com/subscriptions/mysubscription/providers/Microsoft.Authorization/roleDefinitions/180078c9-4b93-45b7-a32f-2093f6126885
--------------------------------------------------------------------------------
RESPONSE 409: 409 Conflict
ERROR CODE: RoleDefinitionWithSameNameExists
--------------------------------------------------------------------------------
{
  "error": {
    "code": "RoleDefinitionWithSameNameExists",
    "message": "A custom role with the same name already exists in this directory. Use a different name."
  }
}
--------------------------------------------------------------------------------
```

Re-attempting eventually worked once everything had replicated so we updated the error messaging in above commit to provide more info regarding the errors and to point out that this can occur with new custom roles and can be re-attempted.

**Comment by abutcher** [comment]
Only e2e failure:

```
: [bz-Unknown][invariant] alert/KubePodNotReady should not be at or above info in all the other namespaces expand_less | 0s

{  KubePodNotReady was at or above info for at least 41m14s on platformidentification.JobType{Release:"4.14", FromRelease:"", Platform:"azure", Architecture:"amd64", Network:"ovn", Topology:"ha"} (maxAllowed=0s): pending for 3m10s, firing for 41m14s:  Jul 09 05:39:15.960 - 2474s W alert/KubePodNotReady ns/openshift-cluster-api pod/capz-controller-manager-67bf7d5bdb-4ggk9 ALERTS{alertname="KubePodNotReady", alertstate="firing", namespace="openshift-cluster-api", pod="capz-controller-manager-67bf7d5bdb-4ggk9", prometheus="openshift-monitoring/k8s", severity="warning"}}
```

/lgtm

**Pull request merged**